clients/notification/src/main/java/org/eclipse/hono/notification/NotificationEventBusSupport.java

```diff
@@ -0,0 +1,161 @@
+     *
+     * @param vertx The vert.x instance to use.
+     * @param notificationType The type of notifications to register the consumer for.
+     * @param consumer The consumer.
+     * @param <T> The class of notifications handled by the given consumer.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     */
+    public static <T extends AbstractNotification> void registerConsumer(final Vertx vertx,
+            final NotificationType<T> notificationType, final Handler<T> consumer) {
+        Objects.requireNonNull(vertx);
+        Objects.requireNonNull(notificationType);
+        Objects.requireNonNull(consumer);
+
+        vertx.eventBus().<NotificationEventBusContainer<T>>consumer(
+                getEventBusAddress(notificationType),
+                msg -> {
+                    final NotificationEventBusContainer<T> notificationEventBusContainer = msg.body();
+                    if (notificationEventBusContainer != null
+                            && notificationType.equals(notificationEventBusContainer.getNotification().getType())) {
+                        consumer.handle(notificationEventBusContainer.getNotification());
+                    } else {
+                        LOG.warn("got event bus message with unexpected body on address [{}]", getEventBusAddress(notificationType));
+                    }
+                });
+    }
+
+    /**
+     * Sends the given notification via the vert.x event bus.
+     *
+     * @param vertx The vert.x instance to use.
+     * @param notification  The notification to send.
+     * @param <T> The class of the given notification.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     */
+    public static <T extends AbstractNotification> void sendNotification(final Vertx vertx, final T notification) {
+        Objects.requireNonNull(vertx);
+        Objects.requireNonNull(notification);
+
+        if (VERTX_INSTANCES_WITH_REGISTERED_CODEC_MAP.putIfAbsent(vertx, true) == null) {
+            vertx.eventBus().registerCodec(new NotificationEventBusContainerLocalCodec());
+        }
+
+        final DeliveryOptions options = new DeliveryOptions();
+        options.setCodecName(NOTIFICATION_EVENT_BUS_CONTAINER_LOCAL_CODEC);
+        options.setLocalOnly(true);
+        vertx.eventBus().publish(
+                getEventBusAddress(notification.getType()),
+                new NotificationEventBusContainer<>(notification),
+                options);
+    }
+
+    private static <T extends AbstractNotification> String getEventBusAddress(final NotificationType<T> notificationType) {
+        return Constants.EVENT_BUS_ADDRESS_NOTIFICATION_PREFIX + "." + notificationType.getTypeName();
+    }
+
+    /**
+     * Container for notification objects to be sent via the vert.x event bus.
+     * @param <T> The notification object class.
+     */
+    static class NotificationEventBusContainer<T extends AbstractNotification> {
```

Comment:
There is a codec needed for custom objects sent via the vert.x event bus. Instead of defining a codec for each `AbstractNotification` sub class, the notification object gets put into the `NotificationEventBusContainer` container object, so that only a codec for `NotificationEventBusContainer` is needed.

Reply:
can't we define a codec for `AbstractNotification`?

Reply:
When passing the codec name in the delivery options on `eventBus.publish()`, the codec class can actually also be a super-class. I've adapted the PR accordingly.